scipy/optimize/tests/test_minpack.py

```diff
@@ -804,6 +804,28 @@ def test_maxfev_and_bounds(self):
         assert_allclose(popt1, 2, atol=1e-14)
         assert_allclose(popt2, 2, atol=1e-14)
 
+    def test_curvefit_omitnan(self):
```

Comment:
It looks like this tests whether the answer with NaNs omitted is approximately equal to the answer with all the original data for the case of 1D `sigma`.

Ideally, we'd test whether the answer with NaNs omitted is *correct* for *all* cases of `sigma`: 0D, 1D, and 2D.

We can test whether the answer is correct by omitting the NaNs manually, maybe using an approach other than a logical mask.

I can't test this code right now, but I'm thinking something like:
```python3
i_x = rng.integers(N, size=5)
i_y = rng.integers(N, size=5)

x[i_x] = np.nan
y[i_y] = np.nan
res = curve_fit(exponential, x, y, p0=p0, sigma=sigma, nan_policy="omit")

i_delete =np.unique(np.concat((i_x, i_y)))
x = np.delete(x, i_delete)
y = np.delete(y, i_delete) 
if sigma.ndim == 1:
    sigma = np.delete(sigma, i_delete)
elif sigma.ndim == 2:
    sigma = np.delete(sigma, i_delete, axis=0)
    sigma = np.delete(sigma, i_delete, axis=1)
ref = curve_fit(exponential, x, y, p0=p0, sigma=sigma)

assert_equal(res, ref)
```

We can parametrize the test (`pytest.mark.parametrize('sigma', ...)`) over the different possibilities of `sigma`.

Does that look like a stronger test to you, @j-bowhay @lucascolley?

Reply:
Yes, this looks like a good way to handle different sigma dimensions. I am on it.